query-engine/core/src/interactive_transactions/transaction.rs

```diff
@@ -0,0 +1,253 @@
+#![allow(unsafe_code)]
+
+use std::pin::Pin;
+
+use crate::{
+    execute_many_operations, execute_single_operation, CoreError, Operation, ResponseData, TransactionError, TxId,
+};
+use connector::{Connection, Transaction};
+use crosstarget_utils::time::ElapsedTimeCounter;
+use schema::QuerySchemaRef;
+use telemetry::helpers::TraceParent;
+use tokio::time::Duration;
+use tracing::Span;
+use tracing_futures::Instrument;
+
+// Note: it's important to maintain the correct state of the transaction throughout execution. If
+// the transaction is ever left in the `Open` state after rollback or commit operations, it means
+// that the corresponding connection will never be returned to the connection pool.
+enum TransactionState {
+    Open {
+        // Note: field order is important here because fields are dropped in the declaration order.
+        // First, we drop the `tx`, which may reference `_conn`. Only after that we drop `_conn`.
+        tx: Box<dyn Transaction>,
+        _conn: Pin<Box<dyn Connection + Send + Sync>>,
+    },
+    Committed,
+    RolledBack,
+    Expired {
+        start_time: ElapsedTimeCounter,
+        timeout: Duration,
+    },
+}
+
+pub enum ClosedTransaction {
+    Committed,
+    RolledBack,
+    Expired {
+        start_time: ElapsedTimeCounter,
+        timeout: Duration,
+    },
+}
+
+impl ClosedTransaction {
+    pub fn error_message_for(&self, operation: &str) -> String {
+        match self {
+            ClosedTransaction::Committed => {
+                format!("A {operation} cannot be executed on a committed transaction")
+            }
+            ClosedTransaction::RolledBack => {
+                format!("A {operation} cannot be executed on a transaction that was rolled back")
+            }
+            ClosedTransaction::Expired { start_time, timeout } => {
+                format!(
+                    "A {operation} cannot be executed on an expired transaction. \
+                     The timeout for this transaction was {} ms, however {} ms passed since the start \
+                     of the transaction. Consider increasing the interactive transaction timeout \
+                     or doing less work in the transaction",
+                    timeout.as_millis(),
+                    start_time.elapsed_time().as_millis(),
+                )
+            }
+        }
+    }
+}
+
+impl TransactionState {
+    async fn start_transaction(
+        conn: Box<dyn Connection + Send + Sync>,
+        isolation_level: Option<String>,
+    ) -> crate::Result<Self> {
+        // Note: This method creates a self-referential struct, which is why we need unsafe. Field
+        // `tx` is referencing field `conn` in the `Self::Open` variant.
+        let mut conn = Box::into_pin(conn);
```

Comment:
TODO: experiment with a dedicate Rust crate that removes `unsafe`, e.g.:
- https://docs.rs/self_cell/latest/self_cell
- https://docs.rs/yoke/latest/yoke/struct.Yoke.html